### DISCUSSION TITLE
accessing element stiffness / Jacobian matrix

### DISCUSSION THREAD

**Discussion by rcontinuum** [idaholab/moose#22394] [Q&A General] [answered]
Hello, 

is it possible to access for each element the corresponding element stiffness / Jacobian matrix in a sort of postprocessing step?  
Say, I implemented a (monolithic) coupled problem with two scalar variables `u` and `v`, developed the corresponding Kernels  and computed in both Kernels
```
Kernel::computeJacobian()
Kernel::computeOffDiagJacobian(const unsigned int jvar_num)   
```
If I use a simple 3-noded triangular C0 element, I expect to end up with an element Jacobian matrix of the size 6x6 (6 = 3 nodes x 2 dofs_per_node). Is this 6x6 matrix somewhere actually computed/stored in MOOSE during assembly? To my understanding `computeJacobian()` and `computeOffDiagJacobian()` computes the diagonal and off-diagonal sub-blocks of the element Jacobian matrix. But do these sub-blocks get actually "assembled" to an (6x6) element Jacobian matrix. 

I tried with a debugger, got a bit lost and ended up in `Assembly.h`
```
std::vector<std::vector<std::vector<DenseMatrix<Number>>>> _sub_Kee;
std::vector<std::vector<std::vector<DenseMatrix<Number>>>> _sub_Keg;
```
My hunch: Do these variables store the Jacobian sub-blocks for each element? And if yes, why do we have two of them? 
(For "special" elements with neighboring contributions etc. many more ...)
 
Background: I would like to compute for each element separate the eigenvalues of the element Jacobian in a kind of postprocessing step. Like: loop over all elements, access the (6x6) element Jacobian matrix, compute its (6) eigenvalues and store the eigenvalues in, say, (6) element aux variables. And, ultimately, store them in an output file.

Maybe, I'm digging to much in the internals of MOOSE which are not intended to be "accessible" to a user. But still, any hint? 

Cheers

rcontinuum

**Comment by GiudGiud** [comment]
Hello

This is where the 'save_in' and 'tags' kernel parameters can be useful.
save_in is on its way to deprecation, but it allows you to save the contribution to the residual and to the diagonal of the Jacobian of any kernel.
tags let you do the same thing but with matrices not just a single vector.
Please look at moose/test/tests/tag or moose/test/tests/misc/save_in for examples of doing this. 

Guillaume

**Reply by loganharbour** [reply]
I would like to add - please don't use save_in. It's deprecated and slated for removal this FY.

**Comment by rcontinuum** [comment]
Hello, 
well, if I understand correct, then 
```
[AuxKernels]
    type = TagMatrixAux
```
and 
```
[Kernels]
  [diff]
    type = Diffusion
    variable = u
    extra_matrix_tags = 'mat_tag1 mat_tag2'
```

gives me via a AuxVariable access to the diagonal entries of the global Jacobian matrix. Like in  
moose/test/tests/tag/2d_diffusion_tag_matrix.i

But what I like to access are the individual element Jacobian matrices. Not the global assembled matrix.  
Say, I have a mesh with 10 elements (3-noded triangles with two scalar variables `u` and `v`). In this case I want 
to access the 10 element stiffness matrices each with a size of 6x6. (6 = 3 nodes x 2 dofs_per_node)

I mean, maybe I could define for each element its own 'block' and for each block an 'extra_matrix_tags'. 
But I would still get only the 6 diagonal entries of each element stiffness matrix. And not the full 6x6 element stiffness 
matrix. And think of a larger mesh and I define for each element its own block. These are a lot of blocks. 

Or do I misunderstand the 'tags' kernel parameter? 

Thanks

**Reply by GiudGiud** [reply]
Hello

I think I misunderstood/misread your question and pointed you to the system jacobian rather than the element jacobians. 

you could make an auxkernel to retrieve a number of these attributes (one component by one component, or use an ArrayAuxKernel) from the Assembly class. Please see it s doxygen. 
https://mooseframework.inl.gov/docs/doxygen/moose/classAssembly.html#a029498f57d8b606ecb049c27cbfaf442

I think you will want to look at _JxW and _jac.

Guillaume

**Reply by rcontinuum** [reply]
Hello @GiudGiud, 
somehow, I overlooked  your comment. 
I will have a look. This class is huge.  

Hmm, I checked also the following methods:

```
void Assembly::prepareJacobianBlock()
Sizes and zeroes the Jacobian blocks used for the current element. 
```
I could try to learn from this the sizes of the Jacobian blocks for current element. 

```
DenseMatrix<Number> & jacobianBlock(unsigned int ivar, unsigned int jvar, TagID tag = 0)
Get local Jacobian block for a pair of variables and a tag. 
```
Does "local Jacobian block" here means "element Jacobian block" here. If yes, 
maybe that is the method to use. Essentially, this method reads from the variable `_sub_Kee`. 

Thanks

**Comment by friedmud** [comment, marked as answer]
Unfortunately.... not really.  If you look in the Assembly object you'll see that we sometimes end up with multiple element jacobians that all get summed together.  There is no place where you can get the final element jacobian.

**Reply by rcontinuum** [reply]
OK, I'll have a look ... Assembly.h is large. So it's like several kernels give contributions to the element Jacobian, but they get separately stored and directly assembled into the global Jacobian. Not first to an element Jacobian which than would get assembled to the global Jacobian. 

Thanks

**Reply by IFFI11Q** [reply]
hello guys ,i need help to compute stiffness and mass matrix , can you help me ?

**Reply by lindsayad** [reply]
What do you need it for? You might want to look at the [Tagging System](https://mooseframework.inl.gov/framework_development/tagging.html)

**Comment by hugary1995** [comment]
@friedmud is right. You can't get the final element Jacobian. But perhaps we can tell you the moose-way of doing what you want, if you tell us what you want to do with the element Jacobian. Maybe you are trying to estimate the CFL condition from element stiffness?

**Comment by rcontinuum** [comment]
>  if you tell us what you want to do with the element Jacobian

I want to compute for each element Jacobian the eigenvalues. Suppose I do a nonlinear (quasistatic Tensor)Mechanics problem. And I know that my initial element stiffness matrix is positive semi-definite. That means, all eigenvalues are non-negative. During deformation the element stiffness matrix may get negative eigenvalues, especially if I have softening material behavior. That's a "poor man's" approach for detecting potential strain localization, "mesh dependent" results etc. Better would be detecting loss of ellipticity on quadrature point level (->acoustic tensor) but that's numerically often very costly.

> @friedmud is right. You can't get the final element Jacobian.

Well, if I could access the contributions of all kernels, i.e., all "sub-blocks" of the element Jacobian for each element, I could do the (local) element assembly "somewhere" by myself. I would also need the "element dof indices" of the all sub-blocks. 

My hunch is that this information is stored in 
`std::vector<std::vector<std::vector<DenseMatrix<Number>>>> _sub_Kee;  // in Assmbly.h
`

Cheers

**Reply by hugary1995** [reply]
> Suppose I do a nonlinear (quasistatic Tensor)Mechanics problem.

If that's what you are after, you can simply look at the eigenvalues of the tangent stiffness. That is either called `_jacobian_mult` or `_dpk1` depending on which system you are using.

**Reply by hugary1995** [reply]
The caveat is that a fair amount of constitutive models defining `_jacobian_mult` doesn't define the exact tangent, so the eigenvalues will be off, and they may not be a good indication of localization. However, if you use the Lagrangian mechanics system, the tangent `_dpk1` is exact.

**Reply by hugary1995** [reply]
> And I know that my initial element stiffness matrix is positive semi-definite.

I hope that's a typo, cuz if the element stiffness isn't positive definite you may run into convergence issues.

**Reply by hugary1995** [reply]
One last comment, I think there is an important question we should've asked at the very beginning: are you using AD?

**Reply by rcontinuum** [reply]
Hello @hugary1995, 

>  I think there is an important question we should've asked at the very beginning: are you using AD?

I don't use AD kernels. For Mechanics I use TotalLagrangian. I should have the consistent tangent. Hopefully ;)

> I hope that's a typo, cuz if the element stiffness isn't positive definite you may run into convergence issues.

It's not a typo. The *element* stiffness matrix, of type 

$$ K_e = \int_{B_e} \nabla_X N  \circ C \cdot \nabla_X N dV $$ 

with N shape functions, C 4th order tangent operator, definitely is positive semi-definite *).
Assume a mesh with one element. You have to put in 2d/3d  the correct 3/6 Dirichlet BCs (to prevent rigid body motions) to make the resultant global stiffness matrix positive definite. The global stiffness matrix "includes" the Dirichlet BCs. Hint: use a 3-noded triangle, lin elast material, you get a 6x6 element stiffness matrix. This 6x6 matrix has rank 3. Use, e.g., Python or Matlab to compute the rank. If you are into Solid Mechanics: you have to put, e.g., a fixed support on one node and a roller on another. 
*) Well, if C is evolving during deformation, the element stiffness matrix can have also negative eigenvalues. And that I want to check numerically, as simple local indicator of "strain localization".

> If that's what you are after, you can simply look at the eigenvalues of the tangent stiffness. 

I'm not interested in the eigenvalues of the global (tangent) stiffness matrix. No, I want to compute the eigenvalues of each element tangent stiffness matrix, separately.

> That is either called _jacobian_mult or _dpk1 depending on which system you are using.

Do you mean

```
const MaterialProperty<RankFourTensor> & _dpk1;  // TotalLagrangianStressDivergenceBase.h
const MaterialProperty<RankFourTensor> & _Jacobian_mult; // StressDivergenceTensors.h

```
These are the 4th order tangent operators. My C in the equation above. But I'm seeking K_e for each element.
Sure, I can implement my equation form above to get the element stiffness matrix. 
But assume I couple Mechanics to temperature, electric field, phase field Kernels. Then the K_e not only contains C. There are also  temperature, ... blocks in K_e and OffDiagonal contributions for coupling. 
Also in this case I'm able to program my own function for the element stiffness matrix. I do these things in my own toy FEM code. But I would basically redo with such a function what MOOSE anyway does. It has, as far as I understand, the sub_blocks of all Kernels contributing to the element Jacobian. Isn't it better (easier?) to use these sub_blocks and assemble them in a function to the "real/full" element stiffness matrix 
instead of implementing my own K_e function. That's the goal off my question.

Cheers and thanks for input.

**Reply by hugary1995** [reply]
okay, I understand your question now (finally...). Thanks for the explanation.

> Sure, I can implement my equation form above to get the element stiffness matrix.
> But assume I couple Mechanics to temperature, electric field, phase field Kernels. Then the K_e not only contains C. There are also temperature, ... blocks in K_e and OffDiagonal contributions for coupling.

That is true. I think you were on the right track to call `jacobianBlock`. You just need to figure out the variable numbers of the variables you are interested in.
